### PR TITLE
feat(Webhook): add HMAC validation to twilio

### PR DESCRIPTION
**Pull request by Oleksiivanov**
## Description

 added  HMAC validation for twilio connector, updated Twilio template, change version parent to 0.11.0 SNAPSHOT, test will pass after https://github.com/camunda/connectors-bundle/pull/746
## Related issues
https://github.com/camunda/connectors-bundle/pull/746

closes #
https://github.com/camunda/connectors-bundle/issues/551

### CODE REVIEW

File: connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/inbound/webhook/model/HttpServletRequestWebhookProcessingPayload.java [hunk 1]
```diff
@@ -35,12 +36,18 @@ public HttpServletRequestWebhookProcessingPayload(
       final Map<String, String> params,
       final Map<String, String> headers,
       byte[] bodyAsByteArray) {
+    this.requestURL = httpServletRequest.getRequestURL().toString();
```
Comment (sbuettner) [inline connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/inbound/webhook/model/HttpServletRequestWebhookProcessingPayload.java:39]: 👍

File: connectors/twilio/element-templates/twilio-webhook-connector.json [hunk 2]
```diff
@@ -60,19 +60,76 @@
...
+        "property": "shouldValidateHmac",
+        "equals": "enabled"
+      }
+    },
+    {
+      "label": "HMAC header",
+      "description": "Name of header attribute that will contain the HMAC value",
+      "type": "String",
+      "group": "endpoint",
+      "optional": true,
```
Comment (sbuettner) [inline connectors/twilio/element-templates/twilio-webhook-connector.json:104]: Shouldnt that be hidden and set to `"x-twilio-signature"`?

File: connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java [hunk 3]
```diff
@@ -101,9 +109,16 @@ private boolean webhookSignatureIsValid(
     final String shouldValidateHmac =
         Optional.ofNullable(context.getShouldValidateHmac()).orElse(disabled.name());
     if (enabled.name().equals(shouldValidateHmac)) {
+      byte[] signatureData;
+      if (extractHeaderIgnoreCase("x-twilio-signature", payload.headers()) != null) {
+        signatureData = extractTwilioSignatureData(payload);
+      } else {
+        signatureData = payload.rawBody();
+      }
+
```
Comment (sbuettner) [inline connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java:118]: Could you elaborate why we dont just use the `hmacHeader` here?

File: connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java [hunk 4]
```diff
@@ -113,6 +128,38 @@ private boolean webhookSignatureIsValid(
...
+          payload.requestURL()
+              + "?"
+              + String.join(
+                  "&",
+                  payload.params().entrySet().stream()
+                      .map(
+                          (Map.Entry<String, String> entry) ->
+                              String.format("%s=%s", entry.getKey(), entry.getValue()))
+                      .toArray(String[]::new));
+      return url.getBytes();
```
Comment (sbuettner) [inline connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java:145]: Can be extracted into static util method.

File: connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java [hunk 5]
```diff
@@ -113,6 +128,38 @@ private boolean webhookSignatureIsValid(
...
+      StringBuilder builder = new StringBuilder(payload.requestURL());
+      List<String> sortedKeys = new ArrayList(signatureData.keySet());
+      Collections.sort(sortedKeys);
+
+      for (String key : sortedKeys) {
+        builder.append(key);
+        String value = signatureData.get(key);
+        builder.append(value == null ? "" : value);
+      }
+      return builder.toString().getBytes();
```
Comment (sbuettner) [inline connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java:159]: Please extract into dedicated static helper melpher method.

File: connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java [hunk 6]
```diff
@@ -113,6 +128,38 @@ private boolean webhookSignatureIsValid(
     return true;
   }
 
+  private byte[] extractTwilioSignatureData(final WebhookProcessingPayload payload)
+      throws IOException {
+
+    if (payload.method().equalsIgnoreCase("GET")) {
```
Comment (sbuettner) [inline connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java:134]: If `hmac` is set we can always extract the data from the url. No need to have a custom twilio logic here.

File: connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/inbound/webhook/model/HttpServletRequestWebhookProcessingPayload.java (hunk 1 above)
Comment (sbuettner) [reply connector-runtime/connector-runtime-spring/src/main/java/io/camunda/connector/runtime/inbound/webhook/model/HttpServletRequestWebhookProcessingPayload.java:39]: We could also just get the whole URL including parameters and put into a type `URL`. Then we dont have to add the parameters again afterwards.

File: connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java [hunk 7]
```diff
@@ -64,55 +56,52 @@ public WebhookProcessingResult triggerWebhook(WebhookProcessingPayload payload)
...
+    if (shouldValidateHmac(context)) {
+      byte[] signatureData;
+      if (payload.method().equalsIgnoreCase("GET")) {
+        signatureData = HttpWebhookUtil.extractSignatureDataFromParams(payload);
+      } else {
+        if (context.getHmacHeader().equalsIgnoreCase("x-twilio-signature")) {
+          signatureData = HttpWebhookUtil.extractTwilioSignatureData(payload);
+        } else {
+          signatureData = payload.rawBody();
+        }
```
Comment (sbuettner) [inline connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java:79]: We could have a feel expression that defines how we calculate the `hmac` based on a list like 
```
=["url", "body"]
```

This way we would get rid of the twilio specific part: 
```
context.getHmacHeader().equalsIgnoreCase("x-twilio-signature")
```

For other connectors we would just define it as `=["body"]` and in the case of `GET`: `=["url", "params"]`

File: connectors/twilio/element-templates/twilio-webhook-connector.json [hunk 8]
```diff
@@ -38,6 +38,32 @@
...
+          "value": "any"
+        },
+        {
+          "name": "Get",
+          "value": "get"
+        },
+        {
+          "name": "Post",
+          "value": "post"
+        }
```
Comment (sbuettner) [inline connectors/twilio/element-templates/twilio-webhook-connector.json:60]: Great idea to limit it here! 👍

File: connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java (hunk 7 above)
Comment (sbuettner) [reply connectors/webhook/src/main/java/io/camunda/connector/inbound/HttpWebhookExecutable.java]: Really cool, thanks @Oleksiivanov

File: connectors/webhook/src/test/java/io/camunda/connector/inbound/signature/HMACTwilioSignatureTest.java [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+  protected static Stream<WebhookProcessingPayloadTest> loadTestCasesFromResourceFile(
+      final String fileWithTestCasesUri) throws IOException {
+    final String cases = readString(new File(fileWithTestCasesUri).toPath(), UTF_8);
+    final ObjectMapper objectMapper = ObjectMapperSupplier.getMapperInstance();
+    return objectMapper
+        .readValue(cases, new TypeReference<List<WebhookProcessingPayloadTest>>() {})
+        .stream();
+  }
+
+  public static class WebhookProcessingPayloadTest implements WebhookProcessingPayload {
```
Comment (sbuettner) [inline connectors/webhook/src/test/java/io/camunda/connector/inbound/signature/HMACTwilioSignatureTest.java:104]: Could be a mock?

**Review by sbuettner — APPROVED** [review]
Great work @Oleksiivanov!

**Review by igpetrov — APPROVED** [review]

**Pull request merged**